src/ci/docker/host-x86_64/dist-various-2/build-fuchsia-toolchain.sh

```diff
@@ -3,52 +3,68 @@
-
-mkdir -p zircon
-pushd zircon > /dev/null
-
-# Download sources
-git init
-git remote add origin https://github.com/rust-lang-nursery/mirror-google-fuchsia-zircon
-git fetch --depth=1 origin $ZIRCON
-git reset --hard FETCH_HEAD
-
-# Download toolchain
-./scripts/download-toolchain
-chmod -R a+rx prebuilt/downloads/clang+llvm-x86_64-linux
-cp -a prebuilt/downloads/clang+llvm-x86_64-linux/. /usr/local
-
-build() {
-  local arch="$1"
-
-  case "${arch}" in
-    x86_64) tgt="zircon-pc-x86-64" ;;
-    aarch64) tgt="zircon-qemu-arm64" ;;
-  esac
-
-  hide_output make -j$(getconf _NPROCESSORS_ONLN) $tgt
-  dst=/usr/local/${arch}-fuchsia
-  mkdir -p $dst
-  cp -a build-${tgt}/sysroot/include $dst/
-  cp -a build-${tgt}/sysroot/lib $dst/
+FUCHSIA_SDK=4xjxrGUrDbQ6_zJwj6cDN1IbWsWV5aCQXC_zO_Hu0XkC
+FUCHSIA_SDK_SHA256=e318f1ac652b0db43aff32708fa70337521b5ac595e5a0905c2ff33bf1eed179
+
+CLANG_ZIRCON=e9a26dbc70d631029f8ee9763103910b7e3a2fe1
+
+install_clang() {
+  # Other dists currently depend on this Clang... keeping only for creation and
+  #  linking of Clang
+  mkdir -p clang_build
+  pushd clang_build > /dev/null
+
+  # Download sources
+  git init
+  git remote add origin https://github.com/rust-lang-nursery/mirror-google-fuchsia-zircon
+  git fetch --depth=1 origin $CLANG_ZIRCON
+  git reset --hard FETCH_HEAD
+
+  # Download clang+llvm
+  ./scripts/download-toolchain
+  chmod -R a+rx prebuilt/downloads/clang+llvm-x86_64-linux
+  cp -a prebuilt/downloads/clang+llvm-x86_64-linux/. /usr/local
+
+  popd > /dev/null
+  rm -rf clang_build
 }
 
-# Build sysroot
-for arch in x86_64 aarch64; do
-  build ${arch}
-done
-
-popd > /dev/null
-rm -rf zircon
-
-for arch in x86_64 aarch64; do
-  for tool in clang clang++; do
-    cat >/usr/local/bin/${arch}-fuchsia-${tool} <<EOF
-#!/bin/sh
-${tool} --target=${arch}-fuchsia --sysroot=/usr/local/${arch}-fuchsia "\$@"
-EOF
-    chmod +x /usr/local/bin/${arch}-fuchsia-${tool}
+install_zircon_libs() {
+  mkdir -p zircon
+  pushd zircon > /dev/null
+
+  # Download Fuchsia SDK (with Zircon libs)
+  curl -LO \
+    "https://chrome-infra-packages.appspot.com/dl/fuchsia/sdk/core/linux-amd64/+/${FUCHSIA_SDK}"
+  echo "$(echo ${FUCHSIA_SDK_SHA256}) ${FUCHSIA_SDK}" | sha256sum --check --status
+  unzip ${FUCHSIA_SDK} "arch/*" -d core-linux-amd64
+
+  for arch in x86_64 aarch64; do
+    mkdir -p /usr/local/${arch}-fuchsia/sysroot/
+    mkdir -p /usr/local/${arch}-fuchsia/lib
+
+    case "$arch" in
+      x86_64) SUB_DIR=x64   ;;
+      *)      SUB_DIR=arm64 ;;
+    esac
+
+    # Necessary sysroot/* and lib directories for build
+    cp -r core-linux-amd64/arch/${SUB_DIR}/sysroot/include /usr/local/${arch}-fuchsia/sysroot
+    cp -r core-linux-amd64/arch/${SUB_DIR}/sysroot/lib /usr/local/${arch}-fuchsia/sysroot
+    cp -r core-linux-amd64/arch/${SUB_DIR}/lib /usr/local/${arch}-fuchsia/
+
+    # Linking to standard clang, clang++, and llvm-ar...
+    # CARGO_TARGET env handles linking to sysroot/* and lib
```

Comment:
This won't work when building C/C++ code. Does this actually work?

If not I think we should keep the script and add the fdio include as we do [when building the rustc used for building Fuchsia](https://fuchsia.googlesource.com/infra/recipes/+/d36a3b608399d38367aead9ae9ede46f65101e61/recipes/contrib/rust_toolchain.resources/generate_config.py#144).

Reply:
This has been working, and I'd like to better understand why it should/shouldn't. Are you saying that for our invocation of Clang, we should link to both the sysroot as well as the fdio package?

Similarly, what is the relationship between the sysroot passed into `CARGO_TARGET_X86_64_FUCHSIA_RUSTFLAGS` vs the sysroot linked directly to our invocation of `clang` and `clang++`?

Reply:
> Similarly, what is the relationship between the sysroot passed into `CARGO_TARGET_X86_64_FUCHSIA_RUSTFLAGS` vs the sysroot linked directly to our invocation of `clang` and `clang++`? Is the first for building Rust code, and the second for building out C/C++?

Yes, if there's a build script that uses the `cc` crate for instance I wouldn't expect it to look at `RUSTFLAGS` to find flags to pass to clang. It would instead look in `CFLAGS` and similar flags. https://github.com/rust-lang/cc-rs#external-configuration-via-environment-variables

A wrapper script is an alternative to putting the flags in the environment.

I guess using the built-in compiler sysroot works for the purpose of building libunwind (and any other C libs that happen to get built). It doesn't really seem _correct_, though; I believe the supported way is to use the SDK sysroot, so I think we should still do that.

> Are you saying that for our invocation of Clang, we should link to both the sysroot as well as the fdio package?

I was, but on second thought let's not do this since it isn't used.

Reply:
This makes tons of sense -- thank you!